src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.h

```diff
@@ -0,0 +1,100 @@
+#include "src/ray/raylet/scheduling/cluster_resource_manager.h"
+
+namespace ray {
+namespace gcs {
+
+class GcsAutoscalerStateManager : public rpc::AutoscalerStateHandler {
+ public:
+  GcsAutoscalerStateManager(ClusterResourceManager &cluster_resource_manager,
+                            GcsResourceManager &gcs_resource_manager,
+                            GcsNodeManager &gcs_node_manager);
+
+  void HandleGetClusterResourceState(
+      rpc::autoscaler::GetClusterResourceStateRequest request,
+      rpc::autoscaler::GetClusterResourceStateReply *reply,
+      rpc::SendReplyCallback send_reply_callback) override;
+
+  void HandleReportAutoscalingState(
+      rpc::autoscaler::ReportAutoscalingStateRequest request,
+      rpc::autoscaler::ReportAutoscalingStateReply *reply,
+      rpc::SendReplyCallback send_reply_callback) override;
+
+  void RecordMetrics() const { throw std::runtime_error("Unimplemented"); }
+
+  std::string DebugString() const { throw std::runtime_error("Unimplemented"); }
+
+ private:
+  int64_t GetNextClusterResourceStateVersion() {
+    return ++cluster_resource_state_version_;
+  }
+
+  /// \brief Get the current cluster resource state.
+  /// \param reply The reply to be filled.
+  ///
+  /// See rpc::autoscaler::GetClusterResourceStateReply::node_states for more details.
+  void GetNodeStates(rpc::autoscaler::GetClusterResourceStateReply *reply);
+
+  /// \brief Get the resource requests state.
+  /// \param reply The reply to be filled.
+  ///
+  /// See rpc::autoscaler::GetClusterResourceStateReply::pending_resource_requests for
+  /// more details.
+  void GetPendingResourceRequests(rpc::autoscaler::GetClusterResourceStateReply *reply);
+
+  /// \brief Get the gang resource requests (e.g. from placement group) state.
+  /// \param reply The reply to be filled.
+  ///
+  /// See rpc::autoscaler::GetClusterResourceStateReply::pending_gang_resource_requests
+  /// for more
+  void GetPendingGangResourceRequests(
+      rpc::autoscaler::GetClusterResourceStateReply *reply);
+
+  /// \brief Get the cluster resource constraints state.
+  /// \param reply The reply to be filled.
+  ///
+  /// See rpc::autoscaler::GetClusterResourceStateReply::cluster_resource_constraints for
+  /// more details. This is requested through autoscaler SDK for request_resources().
+  void GetClusterResourceConstraints(
+      rpc::autoscaler::GetClusterResourceStateReply *reply);
+
+  /// Cluster resources manager that provides cluster resources information.
+  ClusterResourceManager &cluster_resource_manager_;
+
+  /// Gcs node manager that provides node status information.
+  GcsNodeManager &gcs_node_manager_;
+
+  /// GCS resource manager that provides resource demand/load information.
+  GcsResourceManager &gcs_resource_manager_;
+
+  // The default value of the last seen version for the request is 0, which indicates
+  // no version has been seen. When node is added, the version will be incremented.
```

Comment:
> When node is added, the version will be incremented.

Where is the code for that? I think it's incremented for each `HandleGetClusterResourceState` call

Reply:
Ah - i changed the code but haven't updated the comments yet.